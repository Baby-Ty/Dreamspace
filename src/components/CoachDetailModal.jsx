// DoD: no fetch in UI; <400 lines; early return for loading/error; a11y roles/labels; minimal props; data-testid for key nodes.
// CoachDetailModal - Thin wrapper for backward compatibility
// All implementation moved to modular components in ./coach/
// New files:
//   - CoachDetailModal.jsx (main modal shell with tabs & accessibility)
//   - CoachMetrics.jsx (pure presentational metrics display)
//   - TeamMemberList.jsx (pure presentational member list)
//   - CoachingAlerts.jsx (pure presentational alerts)
//   - useCoachDetail.js hook for all modal state & logic

import CoachDetailModal from './coach/CoachDetailModal';

export default CoachDetailModal;
