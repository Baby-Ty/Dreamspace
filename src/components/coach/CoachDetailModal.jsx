// DoD: no fetch in UI; <400 lines; early return for loading/error; a11y roles/labels; minimal props; data-testid for key nodes.
import { useEffect, useRef, useCallback } from 'react';
import { X, BarChart3, Users2, Target, MessageSquare, TrendingUp, Mail, MapPin } from 'lucide-react';
import { useCoachDetail } from './useCoachDetail';
import CoachMetrics from './CoachMetrics';
import TeamMemberList from './TeamMemberList';
import CoachingAlerts from './CoachingAlerts';

/**
 * Modal shell for coach details
 * Handles layout, accessibility, focus trap, and keyboard navigation
 * All business logic delegated to useCoachDetail hook
 * All presentation delegated to child components
 */
export default function CoachDetailModal({ coach, onClose }) {
  const modalRef = useRef(null);
  const closeButtonRef = useRef(null);

  const {
    activeTab,
    setActiveTab,
    filterStatus,
    setFilterStatus,
    sortBy,
    setSortBy,
    teamMetrics,
    coachingAlerts,
    filteredAndSortedMembers,
    summaryMetrics,
    getStatusColor,
    getStatusText
  } = useCoachDetail(coach);

  // Focus management
  useEffect(() => {
    // Focus close button on mount
    if (closeButtonRef.current) {
      closeButtonRef.current.focus();
    }

    // Store previously focused element
    const previouslyFocusedElement = document.activeElement;

    return () => {
      // Restore focus on unmount
      if (previouslyFocusedElement && previouslyFocusedElement.focus) {
        previouslyFocusedElement.focus();
      }
    };
  }, []);

  // Focus trap
  useEffect(() => {
    const handleKeyDown = (e) => {
      if (e.key === 'Tab') {
        const modal = modalRef.current;
        if (!modal) return;

        const focusableElements = modal.querySelectorAll(
          'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
        );
        const firstElement = focusableElements[0];
        const lastElement = focusableElements[focusableElements.length - 1];

        if (e.shiftKey) {
          if (document.activeElement === firstElement) {
            lastElement.focus();
            e.preventDefault();
          }
        } else {
          if (document.activeElement === lastElement) {
            firstElement.focus();
            e.preventDefault();
          }
        }
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    return () => document.removeEventListener('keydown', handleKeyDown);
  }, []);

  // Escape key handler
  const handleEscapeKey = useCallback((e) => {
    if (e.key === 'Escape') {
      onClose();
    }
  }, [onClose]);

  useEffect(() => {
    document.addEventListener('keydown', handleEscapeKey);
    return () => document.removeEventListener('keydown', handleEscapeKey);
  }, [handleEscapeKey]);

  // Prevent body scroll when modal is open
  useEffect(() => {
    const originalOverflow = document.body.style.overflow;
    document.body.style.overflow = 'hidden';
    return () => {
      document.body.style.overflow = originalOverflow;
    };
  }, []);

  // Early return if no coach or team metrics
  if (!coach || !teamMetrics) {
    console.log('ðŸš« CoachDetailModal: No coach or teamMetrics', { 
      hasCoach: !!coach, 
      hasTeamMetrics: !!teamMetrics 
    });
    return null;
  }

  const tabs = [
    { id: 'overview', name: 'Overview', icon: BarChart3 },
    { id: 'team-members', name: 'Team Members', icon: Users2 },
    { id: 'performance', name: 'Performance', icon: TrendingUp },
  ];

  const isActiveTab = (tabId) => activeTab === tabId;

  return (
    <div
      className="fixed inset-0 bg-black bg-opacity-50 backdrop-blur-sm flex items-center justify-center p-4 z-[1000]"
      role="dialog"
      aria-modal="true"
      aria-labelledby="coach-modal-title"
      onClick={onClose}
    >
      <div
        ref={modalRef}
        className="relative bg-white rounded-2xl border border-professional-gray-200 shadow-2xl w-full max-w-7xl max-h-[95vh] overflow-hidden"
        onClick={(e) => e.stopPropagation()}
      >
        {/* Header */}
        <div className="bg-gradient-to-r from-netsurit-red to-netsurit-coral p-4 sm:p-5">
          <div className="flex flex-col sm:flex-row sm:items-center sm:justify-between gap-4">
            <div className="flex items-center space-x-3 sm:space-x-6">
              <img
                src={coach.avatar}
                alt={`${coach.name}'s profile`}
                className="w-16 h-16 sm:w-20 sm:h-20 rounded-full border-4 border-white object-cover flex-shrink-0"
                onError={(e) => {
                  e.target.src = `https://ui-avatars.com/api/?name=${encodeURIComponent(coach.name)}&background=EC4B5C&color=fff&size=100`;
                }}
              />
              <div className="text-white min-w-0">
                <h1 
                  id="coach-modal-title"
                  className="text-xl sm:text-3xl font-bold text-white"
                >
                  {coach.name}
                </h1>
                <p className="text-base sm:text-xl text-white/80 mb-2">Dream Coach</p>
                <div className="flex flex-col sm:flex-row sm:items-center gap-2 sm:gap-4 text-white/80">
                  <div className="flex items-center">
                    <Mail className="w-4 h-4 mr-1" aria-hidden="true" />
                    <span className="text-sm">{coach.email}</span>
                  </div>
                  <div className="flex items-center">
                    <MapPin className="w-4 h-4 mr-1" aria-hidden="true" />
                    <span className="text-sm">{coach.office}</span>
                  </div>
                </div>
                <div className="flex flex-wrap items-center gap-2 mt-3">
                  <span className="bg-white bg-opacity-20 px-3 py-1 rounded-lg text-sm font-medium">
                    Team Size: {summaryMetrics?.teamSize || 0}
                  </span>
                  <span className="bg-white bg-opacity-20 px-3 py-1 rounded-lg text-sm font-medium">
                    Avg Score: {summaryMetrics?.averageScore || 0}
                  </span>
                  <span className="bg-white bg-opacity-20 px-3 py-1 rounded-lg text-sm font-medium">
                    Engagement: {summaryMetrics?.engagementRate || 0}%
                  </span>
                </div>
              </div>
            </div>
            <button
              ref={closeButtonRef}
              onClick={onClose}
              className="self-start sm:self-center p-2 text-white hover:bg-white hover:bg-opacity-20 rounded-xl transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2"
              aria-label="Close coach detail modal"
            >
              <X className="w-6 h-6" />
            </button>
          </div>
        </div>

        {/* Alerts Banner */}
        {coachingAlerts.length > 0 && (
          <div className="bg-netsurit-warm-orange/10 border-b border-netsurit-warm-orange/30 px-6 py-3">
            <div className="flex items-center space-x-2">
              <MessageSquare className="w-5 h-5 text-netsurit-warm-orange" />
              <span className="text-sm font-medium text-netsurit-orange">
                {coachingAlerts.length} team member(s) need attention
              </span>
            </div>
          </div>
        )}

        {/* Tab Navigation */}
        <div className="border-b border-professional-gray-200 bg-professional-gray-50">
          <div className="px-6">
            <nav 
              className="flex space-x-8"
              role="tablist"
              aria-label="Coach detail sections"
            >
              {tabs.map((tab) => {
                const Icon = tab.icon;
                return (
                  <button
                    key={tab.id}
                    onClick={() => setActiveTab(tab.id)}
                    className={`py-4 px-1 border-b-2 font-medium text-sm transition-colors ${
                      isActiveTab(tab.id)
                        ? 'border-netsurit-red text-netsurit-red'
                        : 'border-transparent text-professional-gray-500 hover:text-professional-gray-700 hover:border-professional-gray-300'
                    }`}
                    role="tab"
                    aria-selected={isActiveTab(tab.id)}
                    aria-controls={`${tab.id}-panel`}
                  >
                    <div className="flex items-center space-x-2">
                      <Icon className="w-4 h-4" aria-hidden="true" />
                      <span>{tab.name}</span>
                    </div>
                  </button>
                );
              })}
            </nav>
          </div>
        </div>

        {/* Tab Content */}
        <div className="p-4 sm:p-5 overflow-y-auto max-h-[calc(95vh-300px)]">
          {activeTab === 'overview' && (
            <div
              role="tabpanel"
              id="overview-panel"
              aria-labelledby="overview-tab"
            >
              <div className="space-y-6">
                <CoachMetrics 
                  metrics={summaryMetrics} 
                  coach={coach}
                />
                
                {coachingAlerts.length > 0 && (
                  <CoachingAlerts alerts={coachingAlerts} />
                )}
              </div>
            </div>
          )}

          {activeTab === 'team-members' && (
            <div
              role="tabpanel"
              id="team-members-panel"
              aria-labelledby="team-members-tab"
            >
              <TeamMemberList
                members={filteredAndSortedMembers}
                filterStatus={filterStatus}
                onFilterChange={setFilterStatus}
                sortBy={sortBy}
                onSortChange={setSortBy}
                onViewMember={(member) => console.log('View member:', member)}
                getStatusColor={getStatusColor}
                getStatusText={getStatusText}
              />
            </div>
          )}

          {activeTab === 'performance' && (
            <div
              role="tabpanel"
              id="performance-panel"
              aria-labelledby="performance-tab"
            >
              <CoachMetrics 
                metrics={summaryMetrics} 
                coach={coach}
              />
            </div>
          )}
        </div>
      </div>
    </div>
  );
}

