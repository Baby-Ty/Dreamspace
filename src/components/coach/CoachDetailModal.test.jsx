// DoD: no fetch in UI; <400 lines; early return for loading/error; a11y roles/labels; minimal props; data-testid for key nodes.
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CoachDetailModal from './CoachDetailModal';

// Mock the child components
vi.mock('./CoachMetrics', () => ({
  default: ({ metrics }) => <div data-testid="coach-metrics">Metrics: {metrics?.teamSize}</div>
}));

vi.mock('./TeamMemberList', () => ({
  default: ({ members }) => <div data-testid="team-member-list">Members: {members?.length}</div>
}));

vi.mock('./CoachingAlerts', () => ({
  default: ({ alerts }) => <div data-testid="coaching-alerts">Alerts: {alerts?.length}</div>
}));

// Mock the hook
vi.mock('./useCoachDetail', () => ({
  useCoachDetail: () => ({
    activeTab: 'overview',
    setActiveTab: vi.fn(),
    filterStatus: 'all',
    setFilterStatus: vi.fn(),
    sortBy: 'score',
    setSortBy: vi.fn(),
    teamMetrics: {
      teamSize: 5,
      averageScore: 85,
      engagementRate: 90,
      teamMembers: [
        { id: '1', name: 'Member 1', score: 90 },
        { id: '2', name: 'Member 2', score: 80 }
      ]
    },
    coachingAlerts: [
      { severity: 'high', message: 'Test alert' }
    ],
    filteredAndSortedMembers: [
      { id: '1', name: 'Member 1', score: 90 },
      { id: '2', name: 'Member 2', score: 80 }
    ],
    summaryMetrics: {
      teamSize: 5,
      averageScore: 85,
      engagementRate: 90,
      exceeding: 2,
      onTrack: 2,
      needsAttention: 1
    },
    getStatusColor: () => 'text-professional-gray-700 bg-professional-gray-100',
    getStatusText: () => 'On Track'
  })
}));

describe('CoachDetailModal', () => {
  const mockCoach = {
    id: 'coach-1',
    name: 'John Doe',
    email: 'john@example.com',
    office: 'Cape Town',
    avatar: 'https://example.com/avatar.jpg',
    teamMetrics: {
      teamSize: 5,
      averageScore: 85,
      engagementRate: 90,
      teamMembers: [
        { id: '1', name: 'Member 1', score: 90 },
        { id: '2', name: 'Member 2', score: 80 }
      ]
    },
    alerts: [
      { severity: 'high', message: 'Test alert' }
    ]
  };

  const mockOnClose = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should render modal with role="dialog" and aria-modal', () => {
    render(<CoachDetailModal coach={mockCoach} onClose={mockOnClose} />);
    
    const dialog = screen.getByRole('dialog');
    expect(dialog).toHaveAttribute('aria-modal', 'true');
  });

  it('should display coach name and email', () => {
    render(<CoachDetailModal coach={mockCoach} onClose={mockOnClose} />);
    
    expect(screen.getByText('John Doe')).toBeInTheDocument();
    expect(screen.getByText('john@example.com')).toBeInTheDocument();
  });

  it('should close modal when close button is clicked', async () => {
    const user = userEvent.setup();
    render(<CoachDetailModal coach={mockCoach} onClose={mockOnClose} />);

    const closeButton = screen.getByRole('button', { name: /close coach detail modal/i });
    await user.click(closeButton);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('should close modal when Escape key is pressed', async () => {
    const user = userEvent.setup();
    render(<CoachDetailModal coach={mockCoach} onClose={mockOnClose} />);

    await user.keyboard('{Escape}');

    expect(mockOnClose).toHaveBeenCalled();
  });

  it('should not render when coach is null', () => {
    const { container } = render(<CoachDetailModal coach={null} onClose={mockOnClose} />);
    
    expect(container.firstChild).toBeNull();
  });

  it('should prevent body scroll when modal is open', () => {
    const originalOverflow = document.body.style.overflow;
    
    const { unmount } = render(<CoachDetailModal coach={mockCoach} onClose={mockOnClose} />);

    expect(document.body.style.overflow).toBe('hidden');

    unmount();

    expect(document.body.style.overflow).toBe(originalOverflow);
  });

  it('should match snapshot', () => {
    const { container } = render(<CoachDetailModal coach={mockCoach} onClose={mockOnClose} />);
    
    expect(container.firstChild).toMatchSnapshot();
  });
});
