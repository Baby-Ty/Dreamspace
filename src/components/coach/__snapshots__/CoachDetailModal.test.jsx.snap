// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoachDetailModal > should match snapshot 1`] = `
<div
  aria-labelledby="coach-modal-title"
  aria-modal="true"
  class="fixed inset-0 bg-black bg-opacity-50 backdrop-blur-sm flex items-center justify-center p-4 z-[1000]"
  role="dialog"
>
  <div
    class="relative bg-white rounded-2xl border border-professional-gray-200 shadow-2xl w-full max-w-7xl max-h-[95vh] overflow-hidden"
  >
    <div
      class="bg-gradient-to-r from-netsurit-red to-netsurit-coral p-4 sm:p-5"
    >
      <div
        class="flex flex-col sm:flex-row sm:items-center sm:justify-between gap-4"
      >
        <div
          class="flex items-center space-x-3 sm:space-x-6"
        >
          <img
            alt="John Doe's profile"
            class="w-16 h-16 sm:w-20 sm:h-20 rounded-full border-4 border-white object-cover flex-shrink-0"
            src="https://example.com/avatar.jpg"
          />
          <div
            class="text-white min-w-0"
          >
            <h1
              class="text-xl sm:text-3xl font-bold text-white"
              id="coach-modal-title"
            >
              John Doe
            </h1>
            <p
              class="text-base sm:text-xl text-white/80 mb-2"
            >
              Dream Coach
            </p>
            <div
              class="flex flex-col sm:flex-row sm:items-center gap-2 sm:gap-4 text-white/80"
            >
              <div
                class="flex items-center"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-1"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    height="16"
                    rx="2"
                    width="20"
                    x="2"
                    y="4"
                  />
                  <path
                    d="m22 7-8.97 5.7a1.94 1.94 0 0 1-2.06 0L2 7"
                  />
                </svg>
                <span
                  class="text-sm"
                >
                  john@example.com
                </span>
              </div>
              <div
                class="flex items-center"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 mr-1"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M20 10c0 6-8 12-8 12s-8-6-8-12a8 8 0 0 1 16 0Z"
                  />
                  <circle
                    cx="12"
                    cy="10"
                    r="3"
                  />
                </svg>
                <span
                  class="text-sm"
                >
                  Cape Town
                </span>
              </div>
            </div>
            <div
              class="flex flex-wrap items-center gap-2 mt-3"
            >
              <span
                class="bg-white bg-opacity-20 px-3 py-1 rounded-lg text-sm font-medium"
              >
                Team Size: 
                5
              </span>
              <span
                class="bg-white bg-opacity-20 px-3 py-1 rounded-lg text-sm font-medium"
              >
                Avg Score: 
                85
              </span>
              <span
                class="bg-white bg-opacity-20 px-3 py-1 rounded-lg text-sm font-medium"
              >
                Engagement: 
                90
                %
              </span>
            </div>
          </div>
        </div>
        <button
          aria-label="Close coach detail modal"
          class="self-start sm:self-center p-2 text-white hover:bg-white hover:bg-opacity-20 rounded-xl transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="bg-netsurit-warm-orange/10 border-b border-netsurit-warm-orange/30 px-6 py-3"
    >
      <div
        class="flex items-center space-x-2"
      >
        <svg
          class="w-5 h-5 text-netsurit-warm-orange"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 15a2 2 0 0 1-2 2H7l-4 4V5a2 2 0 0 1 2-2h14a2 2 0 0 1 2 2z"
          />
        </svg>
        <span
          class="text-sm font-medium text-netsurit-orange"
        >
          1
           team member(s) need attention
        </span>
      </div>
    </div>
    <div
      class="border-b border-professional-gray-200 bg-professional-gray-50"
    >
      <div
        class="px-6"
      >
        <nav
          aria-label="Coach detail sections"
          class="flex space-x-8"
          role="tablist"
        >
          <button
            aria-controls="overview-panel"
            aria-selected="true"
            class="py-4 px-1 border-b-2 font-medium text-sm transition-colors border-netsurit-red text-netsurit-red"
            role="tab"
          >
            <div
              class="flex items-center space-x-2"
            >
              <svg
                aria-hidden="true"
                class="w-4 h-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3 3v18h18"
                />
                <path
                  d="M18 17V9"
                />
                <path
                  d="M13 17V5"
                />
                <path
                  d="M8 17v-3"
                />
              </svg>
              <span>
                Overview
              </span>
            </div>
          </button>
          <button
            aria-controls="team-members-panel"
            aria-selected="false"
            class="py-4 px-1 border-b-2 font-medium text-sm transition-colors border-transparent text-professional-gray-500 hover:text-professional-gray-700 hover:border-professional-gray-300"
            role="tab"
          >
            <div
              class="flex items-center space-x-2"
            >
              <svg
                aria-hidden="true"
                class="w-4 h-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14 19a6 6 0 0 0-12 0"
                />
                <circle
                  cx="8"
                  cy="9"
                  r="4"
                />
                <path
                  d="M22 19a6 6 0 0 0-6-6 4 4 0 1 0 0-8"
                />
              </svg>
              <span>
                Team Members
              </span>
            </div>
          </button>
          <button
            aria-controls="performance-panel"
            aria-selected="false"
            class="py-4 px-1 border-b-2 font-medium text-sm transition-colors border-transparent text-professional-gray-500 hover:text-professional-gray-700 hover:border-professional-gray-300"
            role="tab"
          >
            <div
              class="flex items-center space-x-2"
            >
              <svg
                aria-hidden="true"
                class="w-4 h-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="22 7 13.5 15.5 8.5 10.5 2 17"
                />
                <polyline
                  points="16 7 22 7 22 13"
                />
              </svg>
              <span>
                Performance
              </span>
            </div>
          </button>
        </nav>
      </div>
    </div>
    <div
      class="p-4 sm:p-5 overflow-y-auto max-h-[calc(95vh-300px)]"
    >
      <div
        aria-labelledby="overview-tab"
        id="overview-panel"
        role="tabpanel"
      >
        <div
          class="space-y-6"
        >
          <div
            data-testid="coach-metrics"
          >
            Metrics: 
            5
          </div>
          <div
            data-testid="coaching-alerts"
          >
            Alerts: 
            1
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
